File: src/objects/ObjectManager.cpp
#include "../GameServer.h"
#include "GameObject.h"
#include "Vehicle.h"
#include "Bullet.h"
#include "PlayerObject.h"

ObjectManager* ObjectManager::getInstance() {
    GameServer::getInstance()->getObjectManager();
}

#define DX_CREATE_GET_DESC_FUNC(klass, member) \
    klass##Ptr ObjectManager::get##klass(const std::string& label) const { \
        auto it = member.find(label); \
        return it == member.end() ? nullptr : it->second; \
    }

#define DX_CREATE_ADD_DESC_FUNC(klass, member) \
    klass##Ptr ObjectManager::add##klass(const klass##Ptr& obj) {\
        klass##Ptr prev(nullptr); \
        auto prev_it = member.find(obj->getLabel()); \
        if (prev_it != member.end()) {\
            prev = prev_it->second; \
        } \
        member[obj->getLabel()] = obj; \
        return prev; \
    }

#define DX_CREATE_REMOVE_DESC_FUNC(klass, member) \
    klass##Ptr ObjectManager::remove##klass(const klass##Ptr& obj) {\
        klass##Ptr prev(nullptr); \
        auto prev_it = member.find(obj->getLabel()); \
        if (prev_it != member.end()) {\
            prev = prev_it->second; \
            member.erase(prev_it); \
        } \
        return prev; \
    }

#define DX_CREATE_CLEAR_DESC_FUNC(klass, member) \
    void ObjectManager::clear##klass##s() {\
        member.clear(); \
    }

#define DX_CREATE_DESC_FUNCS(klass, member) \
    DX_CREATE_GET_DESC_FUNC(klass, member) \
    DX_CREATE_ADD_DESC_FUNC(klass, member) \
    DX_CREATE_REMOVE_DESC_FUNC(klass, member) \
    DX_CREATE_CLEAR_DESC_FUNC(klass, member)

DX_CREATE_DESC_FUNCS(BulletDesc, m_bulletDescs)
DX_CREATE_DESC_FUNCS(VehicleDesc, m_vehicleDescs)
DX_CREATE_DESC_FUNCS(PlayerDesc, m_playerDescs)
DX_CREATE_DESC_FUNCS(GameObjectDesc, m_objDescs)
DX_CREATE_DESC_FUNCS(ActionDesc, m_actionDescs)

GameObjectPtr ObjectManager::createObject(
    const std::string& key, GameObjectType objType) {
    switch (objType) {
        case GameObjectType::GENERIC: {
            auto desc = getGameObjectDesc(key);
            return desc ? GameObject::createFromDesc(desc) : nullptr;
        }
        case GameObjectType::VEHICLE: {
            auto desc = getVehicleDesc(key);
            return desc ? Vehicle::createFromDesc(desc) : nullptr;
        }
        case GameObjectType::BULLET: {
            auto desc = getBulletDesc(key);
            return desc ? Bullet::createFromDesc(desc) : nullptr;
        }
        case GameObjectType::PLAYER: {
            auto desc = getPlayerDesc(key);
            return desc ? PlayerObject::createFromDesc(desc) : nullptr;
        }
        default:;
    }
    return nullptr;
}

File: src/scripting/auto/api/ObjectManager.lua

--------------------------------
-- @module ObjectManager
-- @parent_module dx

--------------------------------
-- 
-- @function [parent=#ObjectManager] removeGameObjectDesc 
-- @param self
-- @param #std::shared_ptr<GameObjectDesc> obj
-- @return std::shared_ptr<GameObjectDesc>#std::shared_ptr<GameObjectDesc> ret (return value: std::shared_ptr<GameObjectDesc>)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] getGameObjectDesc 
-- @param self
-- @param #string label
-- @return std::shared_ptr<GameObjectDesc>#std::shared_ptr<GameObjectDesc> ret (return value: std::shared_ptr<GameObjectDesc>)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] removeVehicleDesc 
-- @param self
-- @param #std::shared_ptr<VehicleDesc> obj
-- @return std::shared_ptr<VehicleDesc>#std::shared_ptr<VehicleDesc> ret (return value: std::shared_ptr<VehicleDesc>)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] addBulletDesc 
-- @param self
-- @param #std::shared_ptr<BulletDesc> obj
-- @return std::shared_ptr<BulletDesc>#std::shared_ptr<BulletDesc> ret (return value: std::shared_ptr<BulletDesc>)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] clearPlayerDescs 
-- @param self
-- @return ObjectManager#ObjectManager self (return value: ObjectManager)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] getVehicleDesc 
-- @param self
-- @param #string label
-- @return std::shared_ptr<VehicleDesc>#std::shared_ptr<VehicleDesc> ret (return value: std::shared_ptr<VehicleDesc>)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] clearGameObjectDescs 
-- @param self
-- @return ObjectManager#ObjectManager self (return value: ObjectManager)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] createObject 
-- @param self
-- @param #string key
-- @param #int objType
-- @return std::shared_ptr<GameObject>#std::shared_ptr<GameObject> ret (return value: std::shared_ptr<GameObject>)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] clearVehicleDescs 
-- @param self
-- @return ObjectManager#ObjectManager self (return value: ObjectManager)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] removeBulletDesc 
-- @param self
-- @param #std::shared_ptr<BulletDesc> obj
-- @return std::shared_ptr<BulletDesc>#std::shared_ptr<BulletDesc> ret (return value: std::shared_ptr<BulletDesc>)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] addGameObjectDesc 
-- @param self
-- @param #std::shared_ptr<GameObjectDesc> obj
-- @return std::shared_ptr<GameObjectDesc>#std::shared_ptr<GameObjectDesc> ret (return value: std::shared_ptr<GameObjectDesc>)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] removePlayerDesc 
-- @param self
-- @param #std::shared_ptr<PlayerDesc> obj
-- @return std::shared_ptr<PlayerDesc>#std::shared_ptr<PlayerDesc> ret (return value: std::shared_ptr<PlayerDesc>)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] getPlayerDesc 
-- @param self
-- @param #string label
-- @return std::shared_ptr<PlayerDesc>#std::shared_ptr<PlayerDesc> ret (return value: std::shared_ptr<PlayerDesc>)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] addVehicleDesc 
-- @param self
-- @param #std::shared_ptr<VehicleDesc> obj
-- @return std::shared_ptr<VehicleDesc>#std::shared_ptr<VehicleDesc> ret (return value: std::shared_ptr<VehicleDesc>)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] addPlayerDesc 
-- @param self
-- @param #std::shared_ptr<PlayerDesc> obj
-- @return std::shared_ptr<PlayerDesc>#std::shared_ptr<PlayerDesc> ret (return value: std::shared_ptr<PlayerDesc>)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] getBulletDesc 
-- @param self
-- @param #string label
-- @return std::shared_ptr<BulletDesc>#std::shared_ptr<BulletDesc> ret (return value: std::shared_ptr<BulletDesc>)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] clearBulletDescs 
-- @param self
-- @return ObjectManager#ObjectManager self (return value: ObjectManager)
        
--------------------------------
-- 
-- @function [parent=#ObjectManager] getInstance 
-- @param self
-- @return ObjectManager#ObjectManager ret (return value: ObjectManager)
        
return nil
